lib/_stream_readable.js

```diff
@@ -68,11 +68,19 @@ function ReadableState(options, stream) {
   // the point at which it stops calling _read() to fill the buffer
   // Note: 0 is a valid value, means "don't call _read preemptively ever"
   var hwm = options.highWaterMark;
-  var defaultHwm = this.objectMode ? 16 : 16 * 1024;
-  this.highWaterMark = (hwm || hwm === 0) ? hwm : defaultHwm;
+  if (hwm != null) {
+    if (typeof hwm !== 'number') {
+      throw new TypeError('highWaterMark must be a number');
+    }
+
+    if (~~hwm !== hwm) {
+      throw new Error('highWaterMark exceeds valid range');
```

Comment:
This should be a `RangeError`. It might be helpful to say what the valid range actually is.